### ISSUE TITLE
Support requiring Gemfiles inside of Gemfiles (i.e. Gemfileception)

### ISSUE THREAD

**Issue by dtognazzini** [bundler/bundler-features#65]
## Background

Gems, by themselves, only declare dependencies; they do not inform how to satisfy those dependencies. Bundler declares gem dependencies for package management and also specifies how to satisfy the dependencies (e.g. `source`, `path`, `gem(git:)`, etc.)

It would be nice to reuse Gemfile declarations (both dependency specification and satisfaction) in other Gemfiles without copying and pasting.
## Proposal

Provide a `require_gemfile` method in `Bundler::DSL` that takes a path to another Gemfile that can hold common dependency resolution declarations. For example:

_&lt;root&gt;/Gemfile_

``` ruby
require_gemfile "common/Gemfile"
gem "some_gem"
```

_&lt;root&gt;/common/Gemfile_

``` ruby
source "https://my_fav_gemserver.net"
path "gems" do
  gem "my_fav_gem"
end
```

The above would be identical to the following single file:

_&lt;root&gt;/Gemfile_

``` ruby
source "https://my_fav_gemserver.net"
path "common/gems" do
  gem "my_fav_gem"
end
gem "some_gem"
```

**Notice:** the resulting Gemfile nests the relative paths of the _required_ Gemfile (`common/Gemfile`) under the relative path from the _requiring_ Gemfile (`common/`).
## Use Case 1: Sharing common gems

This is very similar to bundler/bundler#3102, which uses `eval_gemfile`. In addition to sharing dependency declaration, `require_gemfile` allows sharing all supported DSL methods. 

What `require_gemfile` offers over `eval_gemfile` is that it rebases relative paths from the required Gemfile to relative paths from the requiring Gemfile. 
## Use Case 2: Sharing source code, repository local gem paths

It's become a common pattern amongst Rails projects to decompose the system into an ecosystem of inline gems containing Rails Engines. The main Rails Application pulls in the inline gems by declaring dependencies in the application-level Gemfile. For example:

_&lt;system_root&gt;/Gemfile_

``` ruby
path "gems" do
  gem "subsystem_one"
  gem "subsystem_two"
  gem "subsystem_three"
end
```

If `subsystem_one` grows crazily, it could be further decomposed into many gems, perhaps:

_&lt;system_root&gt;/subsystem_one/subsystem_one.gemspec_

``` ruby
 s.add_runtime_dependency("subsystem_one-authorization")
 s.add_runtime_dependency("subsystem_one-billing")
 s.add_runtime_dependency("subsystem_one-analytics")
```

_&lt;system_root&gt;/subsystem_one/Gemfile_

``` ruby
# path "gems" here would be <system_root>/subsystem_one/gems
path "gems" do
  gem "subsystem_one-authorization"
  gem "subsystem_one-billing"
  gem "subsystem_one-analytics"
end
```

To support this decomposition, the main application's Gemfile would have to be updated:

_&lt;system_root&gt;/Gemfile_

``` ruby
# Declare dependencies on subsystems
path "gems" do
  gem "subsystem_one"
  gem "subsystem_two"
  gem "subsystem_three"
end

# Add the path to find gems required by subsystem_one
path "subsystem_one/gems"
```

Updating the top level Gemfile for subsystems undergoing structural refactors like this is somewhat painful but reasonably maintainable provided there is only one, conventional "gems" repository for each layer.
### When it becomes unmaintainable

This strategy becomes unmaintainable with one more level of layering.

For example, consider a `super_system` project that pulls in the root of the above project (named here as `system_1`) and the root of another project (`system_2`):

_&lt;super_system&gt;/systems/system_1.gemspec_

``` ruby
 s.add_runtime_dependency("subsystem_one")
 s.add_runtime_dependency("subsystem_two")
 s.add_runtime_dependency("subsystem_three")
```

_&lt;super_system&gt;/Gemfile_

``` ruby
path "systems" do
  gem "system_1"
  gem "system_2"
end
```

The `super_system` project would have to include paths in its Gemfile for the gems required by `system_1`:

_&lt;super_system&gt;/Gemfile_

``` ruby

# Declare dependencies on the systems
path "systems" do
  gem "system_1"
  gem "system_2"
end

# Add paths to find gems required by system_1 and its subsystems
path "systems/system_1/gems"
path "systems/system_1/gems/subsystem_one/gems"
```

The last line couples `super_system` to the internal structure of `system_1`.
### Contrived?

This example may seem a bit contrived, but I work on something similar to this whereby a system is comprised of many Rails Applications. Each Rails Application supplies its own inline gem that the system depends on in its Gemfile. The inline gems provided by the Rails Applications are used by the system to interact with the applications. In the usage above, every restructuring internal to `system_1` would need to be handled by the Gemfile for `super_system`.

With `require_gemfile`, using a convention whereby systems specify their path repositories in a Gempaths file, the above would be simpler:

_&lt;super_system&gt;/Gemfile_

``` ruby
# Reuse gem paths defined by system_1
require_gemfile "systems/system_1/Gempaths"

path "systems" do
  gem "system_1"
  gem "system_2"
end
```

_&lt;super_system&gt;/systems/system_1/Gemfile_

``` ruby
# Use gem paths defined for system_1
require_gemfile "Gempaths"
```

_&lt;super_system&gt;/systems/system_1/Gempaths_

``` ruby
# Reuse gem paths defined by the subsystems
require_gemfile "gems/subsystem_one/Gempaths"
require_gemfile "gems/subsystem_two/Gempaths"
require_gemfile "gems/subsystem_three/Gempaths"
```

_&lt;super_system&gt;/systems/system_1/gems/subsystem_one/Gemfile_

``` ruby
# Use gem paths defined for subsystem_one
require_gemfile "Gempaths"
```

_&lt;super_system&gt;/systems/system_1/gems/subsystem_one/Gempaths_

``` ruby
path "gems"
```

With the above, no layer knows about the internals of the layer beneath it.

**Comment by ccutrer** [comment]
Wow, a long conversation. I completely understand the bundler team's apprehensiveness at making public/supporting a `require_gemfile` method. I just want to present how we use our own "clever hack" of it (which is basically what @indirect wrote for @johnnyshields). I'm also in complete agreement that gemspecs should not specify the source.

We use sub-gemfiles for two important reasons:
1. To break up and maintain the Gemfile. Canvas (https://github.com/instructure/canvas-lms) is a massive Rails project, and depends on 300+ gems (not all of them directly, but many many of them are). To that end, we have a Gemfile.d, and put each bundler group in its on .rb file. This also inherently (partially) solves the use case of developer-specific gems, by just creating your own untracked file in that directory. This wasn't a main goal of creating Gemfile.d, but it's been a happy side effect for some developers. It's also only a partial solution, because developer-specific gems would end up in the lockfile. But that's not a problem, because we don't commit the lock file, because of the next major reason we use sub-gemfiles.
2. Plugins. Canvas is an open source project, but Instructure also maintains proprietary plugins that extend its functionality for us. There are other organizations that also maintain their own Canvas extensions, and by keeping them in plugins, they don't have to worry about merge conflicts keeping the "shared" main Canvas up to date. There are also potential legal benefits, because Canvas is AGPL (being able to keep their own code out of Canvas). Historically, Canvas was a Rails 2 application, and plugins lived in vendor/plugins, and any directory in there automatically had its Gemfile pulled in so that it had its dependency. In Rails 4, vendor/plugins is no longer supported, and we've been migrating both private and public plugins to become real (non-released; referenced by :path) gems, that are simply in the correct location. Either way however, we cannot commit the Gemfile.lock because it would contain references to the proprietary plugins that not everyone has access to. Sadly, this also means we tend to be excessively specific about dependent versions in our actual Gemfile.

Our "clever hack" system has worked well for us, and we're okay maintaining custom monkey-patches of Bundler in our Gemfile when necessary (and removing them if/when they are accepted upstream; https://github.com/bundler/bundler/pull/3171 being one of them). Recently, I ran into an issue that one of our proprietary plugins needs to use a custom fork of a gem we do not maintain. This is not the first time that it has happened, and it's normally easily resolved by using a Gemfile that's not in the public Canvas repo (but included by it), that _either_ points to a private gem server, or points to the git source. Unfortunately, this gem happens to be one that is _also_ used by the main part of Canvas, so having the separate declaration of it in the private Gemfile fragment conflicted with the public Gemfile, which motivated https://github.com/bundler/bundler/pull/3421 (which was rejected, which makes sense because that feature makes no sense in a single-gemfile world; and then I was pointed here)

In conclusion, I'm not pushing to have bundler or rubygems include first class support for either of our scenarios. We would use it if it was available. I'm simply trying to raise awareness that there _are_ use cases out there that would benefit from better support from Bundler, and would be open to any suggestions that are less hacky (especially around not being able to commit the Gemfile.lock).

**Comment by johnnyshields** [comment]
@ccutrer thanks for your awesome comment. I totally get it.

Recently I'm turning my attention towards node.js. Their package manager NPM fully supports the `:git` and `:file` recursive dependency use cases I've outlined above. Granted, NPM has it's own issues, but at least the Node community does not seem to get hung up on these silly semantics of "what" versus "how".

**Comment by indirect** [comment]
@ccruter The entire point of a lock file is to guarantee that the exact same gem versions are available in development and production. What you're describing is the opposite of that: you want to be able to run _different_ versions of the same gems on different machines or different production environments. The closest we can allow to that use-case is using a built gem, and then supplementing one of that gem's dependencies by declaring a git or path gem in the Gemfile. That still produces a single logical outcome—one set of gems for all machines. How do you imagine the lock could possibly work in cases like the one you are asking for?

**Comment by ccutrer** [comment]
I know, and I wish we could do that (natively). Right now we essentially lock down our versions so tight in the Gemfile that the lockfile nearly matches the Gemfile. I think in an ideal world, there would be a lockfile corresponding to each Gemfile that all contribute to the master Gemfile, locking the gems in that specific file (and a final runtime check that none of those lockfiles conflict).

In practice, when we build a new release package, we include the lockfile in it, so we don't have strange problems later if a dependency changed (we also package the gems).

Another not-so-perfect solution is one suggested for per-user gemfile inclusion - having a separate lockfile. Then we could commit the generic lockfile excluding any additional plugins not included in the base project. On deploy, or for developers working with additional plugins, they get a separate lockfile of their own that's not commited (it would be nice if it was "bootstrapped" with the main lockfile, so they're not too far off, but that seems complicated to impossible to keep up to date).

**Comment by johnnyshields** [comment]
@ccutrer supposing `gemspecs` could specify `:git` dependencies, which would be evaluated at higher priority than version numbers `>= 0.3.0` (i.e. Rubygems), does that solve your problem?

**Comment by ccutrer** [comment]
I should mention that in a way we already monkey-patch bundler from our Gemfile to support a separate lockfile - if there is a specific environment variable to enable (experimental) support for Rails 4 in our app, it uses Gemfile.lock4. This allows us to both switch between "normal" and Rails 4 mode very quickly during development (not having to `bundle update` between them, or play some other trick swapping out lockfiles), but also when slowly rolling to production, we have a single release with both lockfiles, and we can A/B test them, or slow roll. Probably not something recommended for everyday use of bundler, though.

**Comment by ccutrer** [comment]
@johnnyshields it would slightly alleviate one of our use cases (i.e. using a Gemfile fragment in addition to a gemspec in order to specify the source of a dependent, non-released gem; which is really not a big deal). It does not solve maintaining a massive Gemfile by breaking it into multiple fragments, or managing a lockfile that may or may not have additional proprietary dependencies.

**Comment by johnnyshields** [comment]
IMHO the reason for the massive Gemfile is that git gem dependencies can't be specified at a per-gem level.

I don't see the point of breaking a large Gemfile into a lot of smaller Gemfiles. In theory you should be breaking your **project** into lots of smaller **gems**, each with it's own **gemspec.** Breaking up the Gemfile alone does nothing to reduce the complexity of your project codebase, your tests take the same amount of time to run, etc.--no real benefit other than purely cosmetic because you prefer to read N files of size M rather than 1 file of size N*M

**Comment by johnnyshields** [comment]
For the lockfile, for the sake of argument let's assume Bundler knows how to read the `:git` and `:file` references from the gemspecs and builds an appropriate lock-file considering those proprietary dependencies.

**Comment by ccutrer** [comment]
This is our "Gemfile", sans anything needed as dependencies of private gems: https://github.com/instructure/canvas-lms/tree/stable/Gemfile.d.  While it would be trimmed down if we weren't explicitly mentioning dependencies-of-dependencies (for locking purposes), it would still be large and unwieldy. Actually, I should probably break rails3 vs. rails4 dependencies into their own files.
Also notice that we do indeed have a multitude of local gems (under the gems folder, and each mentioned by name in the other_stuff.rb part of Gemfile.d).

So, in a nutshell, no, our huge Gemfile (and wanting to break it up) has nothing to do with gem dependencies needing to specify their own dependencies as coming from git source, but a combination of it flat out being a massive project (historically monolithic, but slowly being broken up), and not properly using a lockfile (due to the proprietary plugin problem).

**Comment by johnnyshields** [comment]
OK it looks like you're using some conditional build logic depending on what the user wants, i.e. "postgres on Rails 4" would build one output Gemfile vs. "mysql on Rails 3" -- something like that?

**Comment by ccutrer** [comment]
Kind of. More like "is there stuff in gems/plugins? pull in all the gems there". so non-open sourced stuff just sits there, and you don't have to do anything else - the "master" Gemfile (more specifically, Gemfile.d/plugins.rb) dynamically pulls in the additional gems (or Gemfile.d/~after.rb for legacy plugins in vendor/plugins that aren't gems, but have a Gemfile to declare their dependencies. also support Gemfile fragments from plugins in gems/plugins is in that file, for them to declare the source for non-released gems). Having two layers of dependencies on local gems is actually quite rare, and even in that case is handled by them being siblings, and the master Gemfile automatically pulling in all the local gems.

**Comment by johnnyshields** [comment]
Just for reference, Node.js NPM actually allows you to load two different versions of the same package (=~ gem) at the same time if there are conflicting dependencies, due to the way it's module system works. Rubygems can't do this because it's gems pollute the global namespace with their constants, e.g. attempting to load `ActiveRecord` 4.0 would overwrite `ActiveRecord` 3.0 in memory if you've already loaded it. This sort of thing could be emulated in Ruby but is not the standard, see here for more detailed discussion: http://andrew.ghost.io/emulating-node-js-modules-in-ruby/

**Comment by dtognazzini** [comment]
It's been awhile since I've looked at this and it's quite difficult even for me to distill my exact use cases.

To summarize, in a single Rails project comprised of internal, project-local gems (or Rails Engines), I would use this feature for reusing the following specifications across Gemfiles used by the test environments for each project-local gem:

- `path` specifications to ensure all test environments can find other project-local gems.
- `source` specifications to ensure all test environments use the same version for gems external to the project.
- `group` specifications to share commonly used gems in the Rails' `test` environment.

All of these uses are policy decisions controlled by the project. If/when a project-local gem finds usefulness independent of the project, I would carve out the gem into a new repository and serve via a gem server.

[This discussion](https://github.com/dtognazzini/require_gemfile_example/issues/7) explores one strategy of accomplishing this with `require_gemfile` in more detail.

Still I wonder about [this](#issuecomment-62041050):
>Perhaps a different implementation whereby you could use plain-old require in a Gemfile would be sufficient. Rake supports this by extending the main object with the Rake DSL. Whereas Bundler instance_evals the Gemfile. I wonder if it would be possible to rework some of the internals to allow for using plain-old require.

In that world a `path` specification would be relative to the file it's used in.

**Comment by bbozo** [comment]
Hello :smiley: 

we have a project dealing with 2 rails 4 apps hanging on the same database, a jruby project handling a mobile API and a C ruby CMS of sorts.

To be able to manage this we have extracted the common code into a rails engine (mostly models, some libs) and one day the API server started generating hashes for paperclip in a different way from the CMS server which with S3 means that S3 assets can't be accessed - and after much digging I figure that it's indirectly a freaky dependency issue - an easy mistake to make when you have have 2 Gemfile and 2 Gemfile.lock which share pretty much everything except simple form and choice of rails server.

So I set out to fix the problem once and for all so it doesn't repeat, and OK, I know that you can put the common gems into the gemspec of the engine so I did that. However when I tried that, I realized that those gems don't have git support, local path support and that they don't get required on their own, so it's a bit of a no go, or at least in many ways a step down from what we have now.

Then I google for "how to handle common gem dependencies with engine" and this was the first link, and only link


I was very happy, now I'll to go eat chocolate and think options :disappointed:

**Comment by johnnyshields** [comment]
@bbozo, yes exactly. Welcome to my world. The current Gemfile / gemspec paradigm does not scale well with complex apps. As the Ruby community seems to care little about this problem, I think the best answer is:
* to build a time machine
* go back in time to when you were debating about what framework/language to write your app(s) with
* use node.js

**Comment by bbozo** [comment]
Ι'm starting to get the feeling that bundler source code has gone a bit amok,

basically there's a long line of closed or "low priority" issues that hold back the "split complex rails app into gems/engines" mantra that the Rails team has been pushing a couple years now.

For example:
1. #65 - this issue prevents keeping gem versions in sync among apps using a common engine
2. https://github.com/bundler/bundler/issues/3571 - :path is not supported with "package" - this means that any deployment mechanism depending on bundle package can't work with local gems
3. https://github.com/bundler/bundler/issues/2016 - workaround for last issue is to use git push on the engine and bundle update on the rails apps using it, which is irritating but OK - however when you go this route then every time you want to push a 1-liner code change bundler will try to update ALL of rails dependencies because an engine gem depends on rails. I tried removing rails from the engine gemspec, but it was no use, all of rails dependencies still got pulled

Unfortunately @indirect has been very defensive about all of these issues, according to him:
1. using :git is a hack, you should publish private engines to rubygems.com
2. using :path means "you handle packaging"
3. not updating rails engine dependencies when you want to patch a piece of code from the engine is simply hard and not going to happen

Unfortunately, one gets a feeling that this defensiveness comes from the possible fact that bundler has become to unwieldy to follow Rails development goals and emerging deployment standards,

I get this feeling because in this issue @indirect said that 3000$ isn't enough money for the trouble of developing a feature that should be a priority for the rails community anyway - the ability to handle large projects - so I'm guessing Rails has come to the end of the line where Bundler is concerned and that the rift will continue to grow.

Also, I get a bad feeling from the fact that this issue is the only one to which so far I have seen @indirect help someone circumvent a Bundler design problem, and this was only after he confirmed that @johnnyshields would pay 500$ for a hackish solution.

In fact he says it clearly:

https://github.com/bundler/bundler-features/issues/65#issuecomment-61932167
```(To be super clear to anyone reading this ticket later: the above is a clever but dirty hack, it could break in future versions of Bundler, and the Bundler team only provides help with things like this if you provide monetary compensation.)```

All of this leaves a very bad taste in my mouth because I've personally lost at least 2 weeks of my life handling these issues, and some of these discussions read as an Oracle support worker trying to explain to his clients that what they're seeing is an intended feature and not a serious design issue that needs addressing

Anyway.... my 2 cents

**Comment by bbozo** [comment]
Rails core topic: https://groups.google.com/forum/#!topic/rubyonrails-core/Q0yj6XJEaew

**Comment by segiddins** [comment]
@bbozo I'd ask you to please be polite to André. He has spent thousands of hours on bundler, and he, I, and the rest of the bundler team care greatly about making bundler the best tool it can be.

As for this thread, I think it's gone on long enough. The official position of the bundler team is to reccomend people to publish their gems to their own gem servers, and that reccomendation is unlikely to change without a drastic change in how rubygems or bundler work.

As always, pull requests to bundler are very welcome to improve things you find to be rough edges, but please do keep in mind that were the team stuck maintaining everything in the long run, and to that end we've made the decision that this feature is not one we wish to officially support. Thanks everyone for chiming in!

**Comment by johnnyshields** [comment]
The big problems with problems in my view:

1) Gemspec and gemfile are separate. Gemfile allows using :github/:path, the gemspec does not. Gemfile is for "projects", gemspec is for "gems".
* In NPM (node.js) there is a single file called "package.json" which handles both responsibilities in one scheme.

2) Ruby gems pollute the global namespace. If I two gems that define the namespace `Foobar`, one will monkey patch the other.
* Javascript (ES6) has a module system where each file must `import` and `export`. Thus I can do something like `import Foobar from FoobarGem1 as Foobar` and `import Foobar from FoobarGem2 as Bazqux` to avoid kludges.
* Further point: when NPM gets dependencies, it gets a tree specific to each library. Hence you never get into the situation where two third-party gems clash in Bundler--each gem gets the correct version of it's own dependencies.

Both of these are major obstacles to scaling Ruby in production and are costing me real money in wasted effort. Frankly speaking I much prefer coding in Ruby than Javascript, and Rails is a great framework. But the JS community has it's shit together, the Ruby community does not. IMHO Bundler needs to be scrapped, and Ruby needs a copy of NPM.

(Copying to Ruby Core thread)

**Comment by bbozo** [comment]
@segiddins @indirect please let me know if I can updates parts of my post to not be impolite while still reflecting reality, it is not my intent to insult :heart:

**Comment by johnnyshields** [comment]
I think part of the problem here is that @indirect has done a fantastic job solving the bundler problem under certain assumptions (global namespaces, separation of Gemspec vs. gemfile). If we remove those assumptions however (as NPM has done) a much more scalable solution becomes possible--imagine no more gem conflicts!

While this would certainly be huge undertaking, it's worth noting that Javascript has undergone a similar (painful) transition. Regarding the module (import/export) system for example, there were a variety of competing attempts to solve it via libs--RequireJS, AMD, etc--before it was ultimately incorporated at the language level in EcmaScript 6.

**Comment by johnnyshields** [comment]
@brixen interested to hear your thoughts on this.

**Comment by bf4** [comment]
@johnnyshields @bbozo @dtognazzini There's a lot of words here, and I may be misunderstanding your use case, but you might want to try this in your Gemfile so you can bundle a gem from git or from a local path:  http://www.benjaminfleischer.com/2011/11/01/bundler-ruby-gem-development-tricks/

```ruby
if ENV['APP_GEMS_DIR']
  # for local gem development, use local gem directory so gem changes don't need to be pushed to test
  gem 'my_gem', :path => "#{ENV['APP_GEMS_DIR']}/my_gem", :require => "namespace/my_gem"
else
  gem 'my_gem', :require => "namespace/my_gem", :git => path_to_git_repo, :tag => tagname
end
```

Rubygems isn't going to change to package and install gems (which is what the path/git options in bundler do.  They do the job of gem build, in addition to gem install).

Whenever I've worked on code that started to get crazy with interactions between a gemspec and a Gemfile, I wanted to change the app architecture, not the tools. :)

If you think about it a gemspec's job is to declare dependencies for a library. A Gemfile's job is to declare dependencies for an app.

**Comment by johnnyshields** [comment]
@bf4 please read https://github.com/bundler/bundler-features/issues/65#issuecomment-120257396 carefully.

The problem is not kludging together multiple Gemfiles, and cannot be fixed by a bandaid. There are fundamental design assumptions in Bundler / Rubygems that make scaling Rails beyond "weekend hobby" apps difficult. The more gems in your Gemfile, the more you feel the pain (I'm at 200 and counting)

**Comment by bf4** [comment]
@johnnyshields It's hard to say without seeing your code, but statements like `scaling Rails beyond "weekend hobby" apps difficult.` and `Ruby gems pollute the global namespace.` make it sound like you're missing something.  Plenty of Rails app scale beyond weekend hobby.  And the notion of unintentionally re-opening/re-defining classes in Ruby is one of design, rather than a language issue.  In the old days, Rake had a module called `Task`. So, a Rails app with a `Task` model would interact with Rake in weird ways. Nowadays, the module in Rake would be `Rake::Task`.  That is, good library conventions put the whole library under a namespace.  That's just how it is.

I'm going to stop responding since I'm pretty certain what you want isn't possible without changing the nature of Ruby and rewriting much of its ecosystem.

**Comment by johnnyshields** [comment]
@bf4 have you looked at NPM and ES6 modules?

**Comment by johnnyshields** [comment]
> Ruby gems pollute the global namespace.

They most certainly do. If I have two separate gems which both define a class `Foobar`, then the `Foobar` class gets the methods of both gems. There is no way to load the `Foobar` class as two different names from each gem, short of begging one of the gem authors to rename their class due to name conflict.

**Comment by johnnyshields** [comment]
> In the old days, Rake had a module called Task. So, a Rails app with a Task model would interact with Rake in weird ways.

That's exactly what I mean by "polluting the global namespace." This is a problem not only because of name conflict headaches, but even more so because you can only have a single version of a given gem living in your ecosystem at once. Please study NPM / ES6, which once had the same limitations as Ruby, but has overcome them because the JS community is not content to say "That's just how it is."

**Comment by indirect** [comment]
@bbozo Bundler is both free and open source software. If you don't like how it works, you are welcome to take the source code and make it work the way you want it to instead. Many, many other companies have been able to make Rails engines scale for them without needing this feature. It might be more productive for you to ask them how they did it, rather than demand hundreds or thousands of hours of free software development.

@johnnyshields Bundler is only able to work within the constraints of Ruby and RubyGems. If you want changes to RubyGems, ask the RubyGems team, not the Bundler team. If you want changes to Ruby, ask the Ruby team. If you want to be using NPM instead, please use Node.

Unfortunately, the discussion in this issue has moved beyond a feature request for Bundler, and has turned into demands that gemspecs and even the nature of constants in Ruby itself should be changed. We appreciate the feedback, but we don't accept feature requests for either RubyGems or Ruby here, so I'm going to lock this ticket.
